File: Part 8 - Deep Learning/Section 39 - Artificial Neural Networks (ANN)/ann.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
# Instalando o Keras já instala o Tensorflow e o Theano

# Part 1 - Data Preprocessing

# Importing the dataset
dataset = pd.read_csv("Churn_Modelling.csv")
X = dataset.iloc[:, 3:-1].values
Y = dataset.iloc[:, -1].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X1 = LabelEncoder()
labelencoder_X2 = LabelEncoder()
X[:,1] = labelencoder_X1.fit_transform(X[:,1])
X[:,2] = labelencoder_X2.fit_transform(X[:,2])

# categorical_features já faz referencia que usaremos a coluna de index 0
# esse processo transforma a variável em dummy variables, uma variavel representada por várias colunas, onde cada coluna em 1 representa que esse estado é o ativo
# esse processo deve ser aplicado em variaveis que um estado não assume um valor maior que o outro
# aplicavel á França, Brasil, onde um estado não vale mais que o outro
# não aplicavel á L > M > S, onde um estado vale mais que o outro
onehotencoder = OneHotEncoder(categorical_features=[1])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
# remove uma das colunas das dummy variables
# a bibliotega de regressão linear do python já faz isso, mas algumas bibliotecas não fazem, então é sempre bom fazer
X = X[:,1:] # remove 1 dummy variable de pais
# não precisa fazer essa parte com o genero, pois ele só possui 2 estados (0 e 1)

# Splitting the dataset in train set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Feature scaling
# a biblioteca NÂO aplica feature scalling internamente automaticamente
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
# IMPORTANTE: aqui faz só o transform, pois o fit já foi feito no train, isso faz com que ambos esteja escaladas na mesma base
X_test = sc_X.transform(X_test)
# em classificações o Y não precisa ser escalado. Em regreção sim

# ANN

# Part 2 - Making the ANN

# Importing the Keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

# Initializing the ANN
cla = Sequential()

# Adding the input layer and the first hidden layer
# usa-se na hidden layer a media entre a input layer e a output layer: (11 + 1) / 2 = 6
# output layer será somente 0 ou 1, logo apenas 1 node na output layer
cla.add(Dense(output_dim = 6,
              init = 'uniform',
              activation = "relu",
              input_dim = 11))
# Adding the second hidden layer
cla.add(Dense(output_dim = 6,
              init = 'uniform',
              activation = "relu"))

# Adding the output layer
# se houvesse mais que duas categorias o unit seria atualizado de acordo com o numero
# de dummy variables e a activation seria a softmax (sigmoid para mais que dois estados)
cla.add(Dense(output_dim = 1,
              init = 'uniform',
              activation = "sigmoid"))

# Compiling the ANN
# se houvesse mais que duas categorias a função de loss seria
# categorical_crossentropy 
cla.compile(optimizer = "adam",
            loss = "binary_crossentropy",
            metrics = ["accuracy"])

# Fitting the ANN to the training set
# batch_size é depois de quantas rows que os weights serão atualizados
# é o número de epocas ou seja, o número de vezes que o dataset inteiro será usado
cla.fit(X_train,
        Y_train,
        batch_size = 10,
        epochs = 100)

# Part 3 - Making the predictions and evaluating the model

# Predicting the test set results
Y_prob = cla.predict(X_test)
"""
Y_pred = []
for i in Y_prob:
    if i > 0.5:
        Y_pred.append(1)
    else:
        Y_pred.append(0)
"""
Y_pred = (Y_prob > 0.5)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)

accuracy = (cm[0][0]+cm[1][1])/(cm[0][0]+cm[0][1]+cm[1][0]+cm[1][1])

File: Part 8 - Deep Learning/Section 39 - Artificial Neural Networks (ANN)/ann.R
# Importing the dataset
dataset = read.csv("Churn_Modelling.csv")
dataset = dataset[4:14]

# Encoding the target feature as factor
# decision tre, random forest r naive bayes precisam dessa linha depois trabalham com factores
# dataset$Purchased = factor(dataset$Purchased, levels = c(0, 1))

# Encoding categorical data
# com esse processo a coluna assume valores do tipo factor, e não númericos, apesar de serem números
dataset$Geography = factor(dataset$Geography,
                       levels = c("France","Spain","Germany"),
                       labels = c(1,2,3))
dataset$Gender = factor(dataset$Gender,
                        levels = c("Male","Female"),
                        labels = c(1,2))

# essa biblioteca requere valores númericos e não factores
dataset$Geography = as.numeric(dataset$Geography)
dataset$Gender = as.numeric(dataset$Gender)

# Splitting the dataset in train set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.8)
train_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature scaling
# a biblioteca NÂO aplica feature scalling internamente automaticamente, assim como no python
train_set[,-11] = scale(train_set[,-11])
test_set[,-11]  = scale(test_set[,-11])

# ANN

# Fitting the ANN to the dataset
# essa biblioteca instalada pelo env nao funcionou
# com install.packages('h2o') deu certo
library(h2o)
h2o.init(nthreads = -1)
# dataframe para h2o
# hidden contem o número de nós em cada layer
# Dica: número de nós nas hidden layers é calculado como a media de:
# input nodes = 10
# output nodes = 1 (saida binaria)
# 10+1/2 = 5.5 = 6
cla = h2o.deeplearning(y = "Exited",
                       training_frame = as.h2o(train_set), 
                       activation = "Rectifier",
                       hidden = c(6,6),
                       epochs = 100,
                       train_samples_per_iteration = -2)

# Predicting the Test set results
Y_prob = as.vector(h2o.predict(cla, newdata = as.h2o(test_set[-11])))
Y_pred = (Y_prob > 0.5)

# Making the Confusion Matrix
cm = table(test_set[,11], Y_pred)
accuracy = (1533+193)/2000

# Desconectar do servidor
h2o.shutdown()